From 987b6a1059176a4ebb6ca2578e1cab1731727c97 Mon Sep 17 00:00:00 2001
From: Isuru Fernando <isuruf@gmail.com>
Date: Mon, 25 May 2020 15:19:48 -0500
Subject: [PATCH] Remove flatten-inline-all pass for device without
 autolocals_to_args

In CUDA, we don't convert autolocals to args and use a constant offset
for autolocals. To identify these, they shouldn't be marked as always
inline. This is a hack, but in order to properly fix this, we'll have
to identify the autolocals some other way.
---
 lib/CL/pocl_llvm_wg.cc | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/lib/CL/pocl_llvm_wg.cc b/lib/CL/pocl_llvm_wg.cc
index 561a80ade..f19e019f3 100644
--- a/lib/CL/pocl_llvm_wg.cc
+++ b/lib/CL/pocl_llvm_wg.cc
@@ -229,7 +229,9 @@ kernel_compiler_passes(cl_device_id device, llvm::Module *input,
     passes.push_back("automatic-locals");
 
   if (SPMDDevice) {
-    passes.push_back("flatten-inline-all");
+    if (device->autolocals_to_args) {
+      passes.push_back("flatten-inline-all");
+    }
     passes.push_back("always-inline");
   } else {
     passes.push_back("flatten-globals");
